### ISSUE TITLE
Auto-update private agent fails on Linux when not running as root

### ISSUE THREAD

**Issue by riezebosch** [Microsoft/azure-pipelines-agent#2003]
## Have you tried trouble shooting?
[Trouble shooting doc](https://www.visualstudio.com/en-us/docs/build/troubleshooting)

## Agent Version and Platform
Version of your agent? 2.140.2

OS of the machine running the agent? Linux

## VSTS Type and Version
dev.azure.com

If VisualStudio.com, what is your account name? http://dev.azure.com/raboweb

## What's not working?
```log
[2018-12-06 11:59:46Z INFO SelfUpdater] Finished getting latest agent package at: /home/vsts/agent1/_work/_update.
[2018-12-06 11:59:46Z INFO HostContext] Well known directory 'Bin': '/home/vsts/agent1/bin'
[2018-12-06 11:59:46Z INFO HostContext] Well known directory 'Root': '/home/vsts/agent1'
[2018-12-06 11:59:46Z INFO SelfUpdater] Copy /home/vsts/agent1/_work/_update/bin to /home/vsts/agent1/bin.2.142.1.
[2018-12-06 11:59:46Z INFO HostContext] Well known directory 'Bin': '/home/vsts/agent1/bin'
[2018-12-06 11:59:46Z INFO HostContext] Well known directory 'Root': '/home/vsts/agent1'
[2018-12-06 11:59:46Z INFO SelfUpdater] Copy /home/vsts/agent1/_work/_update/externals to /home/vsts/agent1/externals.2.142.1.
[2018-12-06 11:59:46Z INFO SelfUpdater] Copy any remaining .sh/.cmd files into agent root.
[2018-12-06 11:59:46Z INFO HostContext] Well known directory 'Bin': '/home/vsts/agent1/bin'
[2018-12-06 11:59:46Z INFO HostContext] Well known directory 'Root': '/home/vsts/agent1'
[2018-12-06 11:59:46Z INFO JobDispatcher] Shutting down JobDispatcher. Make sure all WorkerDispatcher has finished.
[2018-12-06 11:59:46Z ERR  Terminal] WRITE ERROR: An error occurred: Access to the path is denied.
[2018-12-06 11:59:46Z ERR  AgentProcess] System.UnauthorizedAccessException: Access to the path is denied. ---> System.IO.IOException: Operation not permitted
   --- End of inner exception stack trace ---
   at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory, Func`2 errorRewriter)
   at Interop.CheckIo(Int64 result, String path, Boolean isDirectory, Func`2 errorRewriter)
   at System.IO.UnixFileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite)
   at System.IO.FileInfo.CopyTo(String destFileName, Boolean overwrite)
   at Microsoft.VisualStudio.Services.Agent.Listener.SelfUpdater.<DownloadLatestAgent>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Agent.Listener.SelfUpdater.<SelfUpdate>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Agent.Listener.Agent.<RunAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.Services.Agent.Listener.Agent.<RunAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.Services.Agent.Listener.Agent.<RunAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Agent.Listener.Agent.<ExecuteCommand>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Agent.Listener.Program.<MainAsync>d__1.MoveNext()
```

## Agent and Worker's Diagnostic Logs
[Agent_20181206-111652-utc (1).log](https://github.com/Microsoft/azure-pipelines-agent/files/2653135/Agent_20181206-111652-utc.1.log)

## My install script
```bash
export HOME=/home/vsts
rm -rf $HOME
mkdir $HOME -p 

# copied from: https://github.com/Microsoft/vsts-agent-docker/blob/master/ubuntu/start.sh#L62-L79
echo Determining matching VSTS agent...
VSTS_AGENT_RESPONSE=$(curl -LsS \
  -u user:$PAT \
  -H 'Accept:application/json;api-version=3.0-preview' \
  "https://$VSTS_ACCOUNT.visualstudio.com/_apis/distributedtask/packages/agent?platform=linux-x64")

if echo "$VSTS_AGENT_RESPONSE" | jq . >/dev/null 2>&1; then
  VSTS_AGENT_URL=$(echo "$VSTS_AGENT_RESPONSE" \
    | jq -r '.value | map([.version.major,.version.minor,.version.patch,.downloadUrl]) | sort | .[length-1] | .[3]')
fi

if [ -z "$VSTS_AGENT_URL" ]; then
  echo 1>&2 error: could not determine a matching VSTS agent - check that account \'$VSTS_ACCOUNT\' is correct and the token is valid for that account
  exit 1
fi

echo Downloading: $VSTS_AGENT_URL
curl -s $VSTS_AGENT_URL > /tmp/agent.tar.gz;

for i in `seq 1 $NR_OF_AGENTS`; do 
    mkdir $HOME/agent$i -p
    
    echo "unpack"
    cd $HOME/agent$i
    tar -zxf /tmp/agent.tar.gz -C .
    chmod -R 777 .
    
    echo "config"
    sudo -u $SUDO_USER \
        ./config.sh \
        --unattended \
        --url https://$VSTS_ACCOUNT.visualstudio.com \
        --auth pat \
        --token $PAT \
        --pool $AGENT_POOL \
        --agent $AGENT_NAME-$hn-$i \
        --work ./_work \
        --acceptTeeEula || true #ignore error

    echo HOME=$HOME >> .env
    
    # copy environment variables into .env file to be picked up by the agent service
    cat /etc/environment | grep -ve PATH | tee -a .env

    echo "install"
    ./svc.sh install $SUDO_USER
    
    echo "start"
    ./svc.sh start
done;

chmod -R 777 $HOME

# make sure custom tools can be installed without root privileges.
chmod -R 777 /opt/hostedtoolcache/
```

The entire VM is based on the [azure pipeline image generation](https://github.com/Microsoft/azure-pipelines-image-generation) repo.

It is deployed in a Virtual Machine Scale Set on Azure.

**Comment by TingluoHuang** [comment]
@riezebosch i think the error means the user doesn't have permission to `/home/vsts/agent1/*.sh`, which seems weird to me.
do you know what is `$SUDO_USER`?

**Comment by riezebosch** [comment]
@TingluoHuang 
1. I would expect that to be fixed with the `chmod -R 777 $HOME` at the end of the install script
2. The same root user that is created while deploying the ARM template

**Comment by riezebosch** [comment]
Any updates on this?

**Comment by riezebosch** [comment]
> @riezebosch i think the error means the user doesn't have permission to `/home/vsts/agent1/*.sh`, which seems weird to me.
> do you know what is `$SUDO_USER`?

I got it resolved, thanks to your reply. After reading it more carefully I focussed on the `*.sh` files and dit a dump of `ls -Rl` on the `$(Agent.HomeDirectory)` and found out there were differences in the permissions between `root` and `raboadmin` (the `$SUDO_USER`).

After I replaced the `chmod` with a `chown` at the bottom of the script the agent was able to update itself.
